04-django/01-python-intro/password_with_constraints.py

```diff
@@ -0,0 +1,18 @@
+import random
+import string
+
+choices = string.ascii_lowercase + string.ascii_uppercase + string.digits +'-'+'_'
+length =random.randint(8, 12)
+def generator():  
+    return [random.choice(choices) for i in range(length)]
+
+def check():
+    good = generator()
+    if ('-'in good or '_'in good):
+        for x in good:
+            if x.isnumeric():return good
+            else: return check()       
+    else: return check()
```

Comment:
I did another version without recursion.

Reply:
Good, that doesn't have the potential to blow up your stack anymore :)

An alternative apprach to generating random passwords until it meets a criteria would be to make different choices wich in sum will meet the criteria (see https://www.erklär-mir-mal.ch/2021-1/09-03-more-on-python.html#/4/0/0 for an example). With this approach, it will be more or less deterministic, how long the function will take...

Reply:
Yes. I did in that way (more or less) when we learned JavaScript. I always want to try some different things, if not, I would never know recursion. Learning is getting interesting (so far). Thank you very much.